lib/taskana-core/src/main/java/pro/taskana/task/internal/ServiceLevelHandler.java

```diff
@@ -82,69 +83,207 @@ public ServiceLevelHandler(
       List<String> requestTaskIds, List<MinimalTaskSummary> existingMinimalTaskSummaries) {
-    BulkOperationResults<String, TaskanaException> bulkLog = new BulkOperationResults<>();
+    BulkLog bulkLog = new BulkLog();
     List<String> nonExistingTaskIds = new ArrayList<>(requestTaskIds);
     List<String> existingTaskIds =
         existingMinimalTaskSummaries.stream()
             .map(MinimalTaskSummary::getTaskId)
             .collect(Collectors.toList());
     nonExistingTaskIds.removeAll(existingTaskIds);
-    for (String taskId : nonExistingTaskIds) {
-      bulkLog.addError(taskId, new TaskNotFoundException(taskId, "Task was not found"));
-    }
+    nonExistingTaskIds.forEach(
+        taskId ->
+            bulkLog.addError(taskId, new TaskNotFoundException(taskId, "Task was not found")));
     return bulkLog;
   }
 
-  private BulkOperationResults<String, TaskanaException> updateAffectedTasks(
-      Instant planned, Map<Duration, List<TaskDuration>> tasksPerDuration) {
-    BulkOperationResults<String, TaskanaException> bulkLog = new BulkOperationResults<>();
-    TaskImpl referenceTask = new TaskImpl();
-    referenceTask.setPlanned(planned);
-    for (Map.Entry<Duration, List<TaskDuration>> entry : tasksPerDuration.entrySet()) {
-      List<String> taskIdsToUpdate =
-          entry.getValue().stream().map(TaskDuration::getTaskId).collect(Collectors.toList());
-      long days = converter.convertWorkingDaysToDays(planned, entry.getKey().toDays());
-      Instant due = planned.plus(Duration.ofDays(days));
-      referenceTask.setDue(due);
-      referenceTask.setModified(Instant.now());
-      long numTasksUpdated = taskMapper.updateTaskDueDates(taskIdsToUpdate, referenceTask);
-      if (numTasksUpdated != taskIdsToUpdate.size()) {
-        bulkLog.addAllErrors(
-            checkResultsOfTasksUpdateAndAddErrorsToBulkLog(
-                taskIdsToUpdate, referenceTask, numTasksUpdated));
+  TaskImpl updatePrioPlannedDueOfTask(
+      TaskImpl newTaskImpl, TaskImpl oldTaskImpl, boolean forRefreshOnClassificationUpdate)
+      throws InvalidArgumentException {
+    boolean onlyPriority = false;
+    if (newTaskImpl.getClassificationSummary() == null
+        || newTaskImpl.getClassificationSummary().getServiceLevel() == null) {
+      onlyPriority = true;
+    }
+
+    if (isPriorityAndDurationAlreadyCorrect(newTaskImpl, oldTaskImpl)) {
+      return newTaskImpl;
+    }
+
+    if (newTaskImpl.getPlanned() == null && newTaskImpl.getDue() == null) {
+      newTaskImpl.setPlanned(Instant.now());
+    }
+
+    DurationPrioHolder durationPrioHolder = determineTaskPrioDuration(newTaskImpl, onlyPriority);
+    newTaskImpl.setPriority(durationPrioHolder.getPriority());
+    if (onlyPriority) {
+      return newTaskImpl;
+    }
+    // classification update
+    if (forRefreshOnClassificationUpdate) {
+      newTaskImpl.setDue(newPlannedDueInstant(newTaskImpl, durationPrioHolder.getDuration(), true));
```

Comment:
Wouldn't it be nicer to move this to the converter?
Like:
newTaskImpl.setDue(converter.addWorkingDaysToTimestamp(newTaskImpl.getPlanned(), durationPrioHolder.getDuration());

What do you think?

Reply:
Good Idea. Actuall, I need 2 methods: addWorkingDaysToInstant and subtractWorkingDaysFromInstant. The reason is, that in the second method I need a negative number of days, but a Duration cannot be negated ...